### PR TITLE
[Merged by Bors] - Improve validator monitor experience for high validator counts

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Myself and others (#3678) have observed  that when running with lots of validators (e.g., 1000s) the cardinality is too much for Prometheus. I've seen Prometheus instances just grind to a halt when we turn the validator monitor on for our testnet validators (we have 10,000s of Goerli validators). Additionally, the debug log volume can get very high with one log per validator, per attestation.

To address this, the `bn --validator-monitor-individual-tracking-threshold <INTEGER>` flag has been added to *disable* per-validator (i.e., non-aggregated) metrics/logging once the validator monitor exceeds the threshold of validators. The default value is `64`, which is a finger-to-the-wind value. I don't actually know the value at which Prometheus starts to become overwhelmed, but I've seen it work with ~64 validators and I've seen it *not* work with 1000s of validators. A default of `64` seems like it will result in a breaking change to users who are running millions of dollars worth of validators whilst resulting in a no-op for low-validator-count users. I'm open to changing this number, though.

Additionally, this PR starts collecting aggregated Prometheus metrics (e.g., total count of head hits across all validators), so that high-validator-count validators still have some interesting metrics. We already had logging for aggregated values, so nothing has been added there.

I've opted to make this a breaking change since it can be rather damaging to your Prometheus instance to accidentally enable the validator monitor with large numbers of validators. I've crashed a Prometheus instance myself and had a report from another user who's done the same thing.

## Additional Info

NA

## Breaking Changes Note

A new label has been added to the validator monitor Prometheus metrics: `total`. This label tracks the aggregated metrics of all validators in the validator monitor (as opposed to each validator being tracking individually using its pubkey as the label).

Additionally, a new flag has been added to the Beacon Node: `--validator-monitor-individual-tracking-threshold`. The default value is `64`, which means that when the validator monitor is tracking more than 64 validators then it will stop tracking per-validator metrics and only track the `all_validators` metric. It will also stop logging per-validator logs and only emit aggregated logs (the exception being that exit and slashing logs are always emitted).

These changes were introduced in #3728 to address issues with untenable Prometheus cardinality and log volume when using the validator monitor with high validator counts (e.g., 1000s of validators). Users with less than 65 validators will see no change in behavior (apart from the added `all_validators` metric). Users with more than 65 validators who wish to maintain the previous behavior can set something like `--validator-monitor-individual-tracking-threshold 999999`.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Related: https://github.com/sigp/lighthouse/issues/3678. I haven't looked closely but maybe this could do what you need @agermain?

**Comment by agermain** [comment]
> Related: #3678. I haven't looked closely but maybe this could do what you need @agermain?

Yup, this is exactly it. I'm happy to contribute as well!

**Comment by paulhauner** [comment]
I'm yet to test this on a running node. I'll do so today and provide and update here when I've done so.

**Comment by paulhauner** [comment]
Alrighty, I've tested this over on a Prater node and ensured that there aren't any individual metrics leaking through with high validator counts. This is ready for review ☺️

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 1]
```diff
@@ -511,21 +581,25 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                        &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ON_CHAIN_INCLUSION_DISTANCE,
-                        &[id],
-                        inclusion_delay as i64,
-                    );
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_int_gauge(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ON_CHAIN_INCLUSION_DISTANCE,
+                            &[label],
+                            inclusion_delay as i64,
+                        );
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: I think using `set_int_gauge` in a loop with `all_validators` might produce some weird results. If different validators have different inclusion distances we'll be overwriting the `all_validators` value with each inclusion distance as we loop through, won't we?

I'm not sure what the best alternative would be, whether we should omit these gauge metrics for `all_validators` or try to compute an aggregate (max, average, etc).

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 2]
```diff
@@ -536,13 +610,19 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                            &[id],
-                            1,
-                        );
+                        self.aggregatable_metric(id, |label| {
+                            metrics::set_int_gauge(
+                                &metrics::VALIDATOR_MONITOR_VALIDATOR_IN_CURRENT_SYNC_COMMITTEE,
+                                &[label],
+                                1,
+                            );
+                        });
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: As above, I think setting the gauge in the loop will toggle the metric for `all_validators` in a strange way. The gauge for `all_validators` should be 1 if _any_ monitored validator is in a sync committee, but instead it'll be set several times and the value from the last validator iterated over will be persisted.

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 3]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    if let Some(delay) = summary.attestation_min_delay {
-                        metrics::observe_timer_vec(
-                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATIONS_MIN_DELAY_SECONDS,
-                            &[id],
-                            delay,
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATIONS_TOTAL,
+                            &[label],
+                            summary.attestations as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: Another gauge, I think we need to sum this one

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 4]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    );
-                    if let Some(distance) = summary.attestation_min_block_inclusion_distance {
+                    self.aggregatable_metric(id, |label| {
                         metrics::set_gauge_vec(
-                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATION_BLOCK_MIN_INCLUSION_DISTANCE,
-                            &[id],
-                            distance.as_u64() as i64,
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATION_AGGREGATE_INCLUSIONS,
+                            &[label],
+                            summary.attestation_aggregate_inclusions as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1501]: Another gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 5]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
                         );
+                    });
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATION_BLOCK_INCLUSIONS,
+                            &[label],
+                            summary.attestation_block_inclusions as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1506]: Another gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 6]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
+                        );
+                    });
+                    if let Some(distance) = summary.attestation_min_block_inclusion_distance {
+                        self.aggregatable_metric(id, |tag| {
+                            metrics::set_gauge_vec(
+                                &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTATION_BLOCK_MIN_INCLUSION_DISTANCE,
+                                &[tag],
+                                distance.as_u64() as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: Gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 7]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    metrics::set_gauge_vec(
-                        &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_COMMITTEE_MESSAGES_TOTAL,
-                        &[id],
-                        summary.sync_committee_messages as i64,
-                    );
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_COMMITTEE_MESSAGES_TOTAL,
+                            &[label],
+                            summary.sync_committee_messages as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: Gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 8]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    metrics::set_gauge_vec(
-                        &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_SIGNATURE_BLOCK_INCLUSIONS,
-                        &[id],
-                        summary.sync_signature_block_inclusions as i64,
-                    );
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_CONTRIBUTION_INCLUSIONS,
+                            &[label],
+                            summary.sync_signature_contribution_inclusions as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1538]: Gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 9]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
+                        );
+                    });
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_SIGNATURE_BLOCK_INCLUSIONS,
+                            &[label],
+                            summary.sync_signature_block_inclusions as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1543]: Gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 10]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    metrics::set_gauge_vec(
-                        &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_CONTRIBUTIONS_TOTAL,
-                        &[id],
-                        summary.sync_contributions as i64,
-                    );
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_CONTRIBUTIONS_TOTAL,
+                            &[label],
+                            summary.sync_contributions as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1554]: Gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 11]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
                     if let Some(delay) = summary.sync_contribution_min_delay {
-                        metrics::observe_timer_vec(
+                        self.aggregatable_metric(id, |tag| {
+                            metrics::observe_timer_vec(
                             &metrics::VALIDATOR_MONITOR_PREV_EPOCH_SYNC_CONTRIBUTION_MIN_DELAY_SECONDS,
-                            &[id],
+                            &[tag],
                             delay,
                         );
+                    });
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: Indentation is funny here

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 12]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    if let Some(delay) = summary.block_min_delay {
-                        metrics::observe_timer_vec(
-                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_BEACON_BLOCKS_MIN_DELAY_SECONDS,
-                            &[id],
-                            delay,
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_BEACON_BLOCKS_TOTAL,
+                            &[label],
+                            summary.blocks as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1572]: Another gauge (sorry)

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 13]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
-                    if let Some(delay) = summary.aggregate_min_delay {
-                        metrics::observe_timer_vec(
-                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_AGGREGATES_MIN_DELAY_SECONDS,
-                            &[id],
-                            delay,
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_AGGREGATES_TOTAL,
+                            &[label],
+                            summary.aggregates as i64,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs:1591]: ... gauge

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 14]
```diff
@@ -1347,125 +1480,161 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
+                            summary.proposer_slashings as i64,
+                        );
+                    });
+                    self.aggregatable_metric(id, |label| {
+                        metrics::set_gauge_vec(
+                            &metrics::VALIDATOR_MONITOR_PREV_EPOCH_ATTESTER_SLASHINGS_TOTAL,
+                            &[label],
+                            summary.attester_slashings as i64,
+                        );
+                    });
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: Three more gauges

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks great on the whole!

I'm afraid I might have misunderstood Prometheus gauges... in case I haven't, there seems to be an issue where gauges set in a loop will overwrite each other. I started flagging them individually before I realised how many there were, so apologies for the log volume ;)

I think some in-house aggregation might be needed for the gauges? Or we could skip them for now and come back for them. Thoughtful aggregation seems like it could be quite a bit of work.

**Comment by paulhauner** [comment]
> there seems to be an issue where gauges set in a loop will overwrite each other

Good call regarding the gauges, there seems to be little/no value in looping through them and re-writing them. I've disabled all gauges unless we're doing individual tracking. We can come back and add averaging in another PR, if we desire. Thanks!

File: beacon_node/beacon_chain/src/validator_monitor.rs [hunk 15]
```diff
@@ -553,17 +633,21 @@ impl<T: EthSpec> ValidatorMonitor<T> {
...
+                            );
+                        }
+                        if self.individual_tracking() {
+                            debug!(
+                                self.log,
+                                "Validator isn't part of the current sync committee";
+                                "epoch" => current_epoch,
+                                "validator" => id,
+                            );
+                        }
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/validator_monitor.rs]: We could collapse these into one `if`, or into the `else` of the parent expression

**Review by michaelsproul — APPROVED** [review]
Awesome! :tada:

One small nit, but happy to merge with or without it

File: beacon_node/beacon_chain/src/validator_monitor.rs (hunk 15 above)
Comment (paulhauner) [reply beacon_node/beacon_chain/src/validator_monitor.rs]: Good suggestion, included in 6eb7dc249

**Comment by arnetheduck** [comment]
fwiw, we've had this feature in nimbus for a while where we use the metric label `total` instead of the pubkey when aggregating - potentially, using this label we'd be a bit more compatible.

**Comment by paulhauner** [comment]
> fwiw, we've had this feature in nimbus for a while where we use the metric label total instead of the pubkey when aggregating - potentially, using this label we'd be a bit more compatible.

Just in the nick of time! I've swapped to `total` so we're more like Nimbus ☺️

**Pull request closed without merging**